### ISSUE TITLE
Adding tensor classes?

### ISSUE THREAD

**Issue by JeanKossaifi** [tensorly/tensorly#67]
Adding tensor classes
================

This is an issue to open to discussion a long standing design decision, namely whether to create subclasses for tensors. 

Specifically there are two design decisions to be made: 
1) whether to create a tensor class that subclasses each of the backend's ndarray structures
2) whether to create classes for the decomposed tensors (e.g. Kruskal and Tucker)

Feel free to leave a comment with your opinions!

----

1- Core tensor class: To inherit or not to inherit?
==================================

We could create a class tensor, that, for each backend, would inherit from that backend's ndarray structure, e.g. 
```python
class Tensor(backend.NDArray):
    ...
```

Pros
-----

This would offer a nice user interface.  We could have fancy syntax such as:
```python
unfolding = tensor.unfold(mode=0)
unfolding.fold()
tensor.mode_dot(matrix)
```

Currently, while folding, one needs to pass along the full-tensor's shape as this isn't stored when unfolding. 

Cons
-----

The main draw back is the added complexity in the code. Specifically, when subclassing the ndarray's class:
* This needs to be done for each backend (with the issue that they don't all work the same. In NumPy for instance, we would need to call __new__ and __array_finalize__).
* Need to redefine all operations (add, mul, etc)
* Potentially breaks when trying to manipulate jointly a tensorly.tensor and an ndarray from the original backend?  We’d have to make sure that all operations return a tensorly.tensor. However, when calling native function from the backend library, these might return ndarrays while the user might expect tensorly.tensors

-----

2- Classes for the decomposed tensors
===========================
This part is less controversial in that in doesn't break anything and doesn't seem to really have many cons.  It would consist in adding classes for the tensors in decomposed form, e.g. KruskalTensor, TuckerTensor, etc -- right now we only have functions that return tuples of native ndarrays from each backend.

a- Pros
-------

* Shorter syntax (e.g. tensor.mode_dot(matrix) VS kruskal_mode_dot(tensor, matrix)
* User needs not think about which function to call 
* Makes it easier to operate directly on decomposed tensors

b- Cons
-------

* More code to maintain, and more complexity in the codebase -- but in that case doesn't seem to be a major issue.
* The big question is **do we still allow simple tuples or do we enforce all decompose tensors to be the right class**? If we enforce, we can simply check with type the decompose tensor is and call the correct method in the functions. For instance, `mode_dot(tensor, matrix, mode)` would work **transparently** on both full tensors and decomposed ones.

c- Possible implementations
---------------------------
The most straightforward would be to subclass tuple or named tuple. This has the advantage of not breaking the current API, users can still directly get e.g. 

```python
core, factors = tucker(tensor, rank)
```
would be as valid as
```python
tucker_tensor_instance = tucker(tensor, rank)
```

**Comment by IlyaKisil** [comment]
Hi. I think, having classes for both ``Tensor`` and ``KruskalTensor`` etc, is more benefitial as it brings even more structure and flexibility for further extentions, that could be implemented under the hood without affecting the core part. On top of that, this will allow to prevent potential mistakes. 
Consider the following scenario:
1. There is a 3d array of shape ``(10, 20, 30)``

2. It is unfolded along mode-0: ``array_2d = unfold(array_3d, mode=0)``. This results in 2d array of shape ``(10, 600)``

3. Then this unfolding is multiplied from the left with 2d array of shape ``(100, 10)`` to produce 2d array of shape ``(100,600)``

4. Now if this output is folded back to 3d array using incorrect values for the shape parameter, i.e 
   ```python
   array_3d_modified = fold(array_2d_modified, mode=0, shape=(10, 40, 150)
   ```
   Then the multidimensional relations between values are broken and resulting 3d array is just a bunch 
   of numbers. This could be easily prevented when additinal infromation about the correct shape is 
   stored within an instance of ``Tensor`` class and as you pointed out would simplify user interface, i.e. 
   ```python
   array_2d_modified.fold(mode=0)
   ```
6. Similarly, function ``context(tensor)``, could become a method or property of ``Tensor`` class.
7. Additionally, autocompletion will come handy whether 

As far as I understood the API you are proposing, for every tensor form (full, kruskal, tucker, tensor-train) there will be 5 classes as of this moment. I agree that this will all cons that you have listed. Have you considered to have a place holder for the data and a reference to which backend to use. This seems more natural and easier to incorporated with the existing code than inheritance. For example, something like:

```python
from tensorly.backend import mxnet_backend, numpy_backend


class Tensor:  
    def __init__ (self, array, backend="numpy"):
       if backend == "numpy":
           self.backend = numpy_backend    
       elif backend == "mxnet":
           self.backend = mxnet_backend
       self.array = self.backed.tensor(array)

    def change_backend(self, new_backend):
       if new_backend == "numpy":
           self.backend = numpy_backend    
       elif new_backend == "mxnet":
           self.backend = mxnet_backend
       self.array = self.backed.tensor(self.array)
    
    def ndim(self):
        return self.backend.ndim(self.array)
   
    def unfold(self, mode):
        return self.backend.unfold(self.array, mode)
    
    ...

```

On one hand, this would allow to have two instances of ``Tensor`` that would to use different backend without manually switching it when needed, i.e. ``tl.set_backend("mxnet")``. On the other hand, these two instances would not be compatible with each other without conversion to the similar data type.

Overall, I belive that exploiting OOP side of python is definetely a benifit for both tensor, kruskal/tucker tensors.

**Comment by JeanKossaifi** [comment]
I agree on the advantages of having a tensor class (in fact one of the first versions of TensorLy before it was open-sourced had such classes). While having classes for decomposed tensors is reasonably straightforward (and transparent for the user), creating a Tensor class isn't as much.
 
Main reasons that make me hesitate:

1) It may over-complicate the code (making it harder to read and maintain). To change things you'd have to go back and forth between classes. Scikit-learn did an amazing job creating a clean and API, and I would like to follow it as much as possible. One of their principles is the *non-proliferation of classes*.. 
2) Related to the previous point, TensorLy algorithms, just like Scikit-Learn, just accept any tensor as input, you don't need to first wrap it in a new class. Practically, you can take any numpy array and apply tensor decomposition to it without having to convert back-and-forth to and from TensorLy. We would have to subclass each backend's core tensor class to maintain this behaviour, which brings us back to point 1.

The issue with the class you wrote is that Tensor *has* an ndarray but it *isn't* one. The following code currently works but wouldn't with such class:
```python
import pytorch
import tensorly as tl
from tensorly.decomposition import tucker, parafac
tl.set_backend('pytorch')

t = torch.tensor(...)
core, factors = tucker(t, ...)

# you can directly use core and factors with any torch function including deep nets.
tt = net(core) # some neural net

# you can apply tensorly function to any pytorch tensor
factors = parafac(tt)
```

With a class like you proposed, we would need additional logic (issue 2), such as
```python
# pytorch tensor to tensorly Tensor
wraped_tensor = tensorly.Tensor(pytorch_tensor)

# for tensorly tensor to be used with pytorch
pytorch_tensor = wraped_tensor.array
```
So while we simplify some actions, we make others more complex, forcing users to keep track what type of object they are manipulating. Some operations are also impossible without additional magic, e.g. `pytorch_tensor + wraped_tensor` will not work without wraping the pytorch tensor under the hood.  

They are also some subtle issuess. For instance, for unfolded tensors, we'd have to keep track of the folded shape, and updated it as necessary. Consider the following:
```python
tensor = random_tensor(size=(10, 10, 10))
matrix = random_tensor(size=(100, 8))

unfolded = tensor.unfold(mode=0)
unfolded = unfolded.dot(matrix)

unfolded.fold() # ???
```
However, it is exactly cases like this (first adding an `unfolded_tensor` class, then dealing with edge cases, etc) that end up quickly making a codebase unmaintanable (issue 1)..

The way forward for this is probably to subclass the backends respective ndarray structures to allow to combine with them easily, and it might be worth just trying, and see where this leads..

**Comment by JeanKossaifi** [comment]
Tentatively added by #87
